### PR TITLE
Conversant s2s adapter

### PR DESCRIPTION
**Pull request by pycnvr**
Conversant prebid server adapter
Supports banner and videos

Maintainer: mediapsr@conversantmedia.com

[Pull request for proper type conversion for parameters](https://github.com/prebid/Prebid.js/pull/1863)

### CODE REVIEW

**Comment by dbemiller** [comment]
hey @pycnvr, thanks for the PR!

I'm willing to review this as is... but wanted to warn you about some upcoming changes for #150.

I opened #196 a few days ago, which includes [a new Bidder interface](https://github.com/prebid/prebid-server/pull/196/files#diff-14b3aba5d2fbcc70f5b2ed53de863330R10) and [better docs for adding new adapters](https://github.com/prebid/prebid-server/pull/196/files#diff-64c48fb391d6a78334bfbc1a86963c60). We're planning to merge this sometime very soon... which will [deprecate the interface you're implementing](https://github.com/prebid/prebid-server/pull/196/files#diff-a58031ae9ef072d17ec94b4011d09707).

The new interface should be much easier to implement than the old one... and we don't plan to add any more features to the old one. Would you still like me to review this, or would you rather wait a few days for us to merge #196, and then re-submit one using the new code?

**Comment by pycnvr** [comment]
@dbemiller Thanks for the info.  Would the old adapter interface continued to be supported?

**Comment by dbemiller** [comment]
The old interface will continue to work with the `/auction` endpoint, but _not_ the `openrtb2/auction` one.

**Comment by pycnvr** [comment]
Will Prebid.js switch to call openrtb/auction instead?  And would there be a new user sync model?

**Comment by mkendall07** [comment]
@pycnvr 
yes prebid.js will switch to the /openrtb endpoint in the future.

**Comment by dbemiller** [comment]
The same user syncs will exist regardless of which endpoint you hit. The overall model is roughly the same too... the only difference will be that `/openrtb2/auction` will return the syncs somewhere inside `bidresponse.ext`, to stay consistent with the spec.

**Comment by pycnvr** [comment]
Thanks, folks.  The new change looks good, and is more consistent with Prebid.js 1.0.  However, would you mind still review this adapter with it's legacy interface anyway?  We will like to adopt the new bidder interface later.  I'll be happy to wait for #196 first and rebase to this branch if that makes this task easier.

**Comment by dbemiller** [comment]
alright, that sounds good. Thanks also for the feedback :). You're the first adapter writer to comment on it specifically... so it's good to get your impressions.

No need to wait for #196. None of those changes affect the current code flow... so this adapter probably won't have to change at all. Once it merges, I'll open issues for existing adapter updates and ping you on the one for this adapter.

Could you paste a sample request which will make your adapter return some bids?

**Comment by pycnvr** [comment]
Yes, I should have included some sample requests.  These are driven thru Prebid.js.  Thanks!

Banner:

```javascript
var adUnits = [{
    code: 'div-gpt-ad-1460505748561-0',
    sizes: [[300, 250], [300,600]],

    // Replace this object to test a new Adapter!
    bids: [{
        bidder: 'conversant',
        params: {
            site_id: '108060'
        }
    }]
}];
```

Video:

```javascript
var videoAdUnit = {
    code: 'video1',
    sizes: [640,480],
    mediaTypes: {
        video: {
            context: "instream",
            mimes: ['video/mp4'],
            protocols: [1, 2],
        }
    },
    bids: [
        {
            bidder: 'conversant',
            params: {
                site_id: '88563',
            }
        }
    ]
};
```

File: adapters/conversant/conversant.go [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+	"golang.org/x/net/context/ctxhttp"
+)
+
+type ConversantAdapter struct {
+	http         *adapters.HTTPAdapter
+	URI          string
+	usersyncInfo *pbs.UsersyncInfo
+}
+
+type FlexBool bool
```
[the remark this answers is not in the document]
Comment (pycnvr) [reply adapters/conversant/conversant.go]: Thanks for a thorough review!

This is indeed for incompatible types.  There is a custom parameter called 'secure' which some pubs use 1 and 0, while others assigns true and false.  Javascript tolerates either choices but this is problem in Go.  Anyway FlexBool is an attempt to keep both kinds of settings working.  It's not ideal for sure.  I'll take a look on the prebidserveradapter side and maybe we can do the conversion there.

File: adapters/conversant/conversant.go [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+	if req.IsDebug {
+		debug.StatusCode = resp.StatusCode
+	}
+
+	if resp.StatusCode == 204 {
+		return nil, nil
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
```
[the remark this answers is not in the document]
Comment (pycnvr) [reply adapters/conversant/conversant.go]: Good call.  Thanks.

**Comment by pycnvr** [comment]
I'll submit a corresponding PR on the prebid.js side to handle the type conversion.  Thanks!

**Comment by pycnvr** [comment]
Sorry there is an error after rbase.  I'll work on it.

**Comment by dbemiller** [comment]
No worries. Could you also leave contact info for the adapter's maintainer somewhere in a comment here?

We don't store these in the code yet, like Prebid.JS does... but plan to start doing that soon. If you leave it in a comment here, I'll add it for you when I make that PR.

**Comment by pycnvr** [comment]
It was missing an entry in the new bidderMap.  All good now.
Please add mediapsr@conversantmedia.com as the maintainer email.  Thank you!

**Comment by dbemiller** [comment]
@pycnvr Not sure if I'm missing something... but I'm not getting bids back through your adapter. I tested in `Prebid.js` master, as well as your branch in https://github.com/prebid/Prebid.js/pull/1863

Both of the screenshots below were taken from `integrationExamples/gpt/prebidServer_example.html` in Prebid.js. I swapped the adUnits you defined here into the test page, and pointed `s2sConfig.endpoint` at my local machine. Debug mode shows the Conversant server returning a `204`, which seems to get returned as no-bid.

For banner:

screen shot 2017-11-27 at 2 28 58 pm

For video:

screen shot 2017-11-27 at 2 30 20 pm

**Comment by dbemiller** [comment]
That's a good guess, and you're right that my original request had an IP of `::1`... but it doesn't look like that's the only problem.

I scraped the request from `Prebid.js` and loaded it into Postman:

```
{
  "account_id":"b1c81a38-1415-42b7-8238-0d2d64016c27",
  "tid":"b18a09a3-adb5-4386-a1af-f7a96cf730df",
  "max_bids":1,"timeout_millis":1000,
  "url":"http://test.mysite.com:9999/integrationExamples/gpt/hello_world.html?pbjs_debug=true",
  "prebid_version":"0.30.0",
  "ad_units": [
    {
      "code": "div-gpt-ad-1460505748561-0",
      "is_top_frame": 0,
      "sizes": [
        {
    	  "w": 300,
    	  "h": 250
    	},
    	{
    	  "w": 300,
    	  "h": 600
    	}
      ],
      "bids": [
        {
          "bidder": "conversant",
          "params": {
            "site_id": "108060"
          }
        }
      ]
    }
  ],
  "is_debug":true
}
```

and set the `X-Forwarded-For` header to `152.193.6.74`. With the debug flag set, I can see that IP reflected in the request it sent to your server:

```
{
    "tid": "b18a09a3-adb5-4386-a1af-f7a96cf730df",
    "status": "no_cookie",
    "bidder_status": [
        {
            "bidder": "conversant",
            "response_time_ms": 48,
            "no_cookie": true,
            "no_bid": true,
            "usersync": {
                "url": "http://prebid-match.dotomi.com/prebid/match?rurl=http%3A%2F%2Flocalhost%3A8000%2Fsetuid%3Fbidder%3Dconversant%26uid%3D",
                "type": "redirect"
            },
            "debug": [
                {
                    "request_uri": "http://media.msg.dotomi.com/s2s/header/24",
                    "request_body": "{\"id\":\"b18a09a3-adb5-4386-a1af-f7a96cf730df\",\"imp\":[{\"id\":\"div-gpt-ad-1460505748561-0\",\"banner\":{\"format\":[{\"w\":300,\"h\":250},{\"w\":300,\"h\":600}],\"w\":300,\"h\":250},\"displaymanager\":\"prebid-s2s\",\"secure\":0}],\"site\":{\"id\":\"108060\",\"domain\":\"othersite.com\",\"page\":\"http://www.othersite.com\"},\"device\":{\"ua\":\"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36\",\"ip\":\"152.193.6.74\"},\"user\":{},\"at\":1,\"tmax\":1000,\"source\":{\"fd\":1,\"tid\":\"b18a09a3-adb5-4386-a1af-f7a96cf730df\"}}",
                    "status_code": 204
                }
            ]
        }
    ]
}
```

but... as you can see, it still gives a 204/no-bid response.

**Comment by pycnvr** [comment]
This one is missing buyeruid, probably due to postman not having the prebid server cookie.

**Comment by dbemiller** [comment]
That got it, thanks. 

Just FYI: PBS doesn't guarantee that that buyeruid exists. It'll only exist if the PBS host's domain has synced with your servers recently.

If a buyeruid is a requirement for bidding, you should implement `SkipNoCookies()` to return `true`. PBS core won't even call your adapter if it can't find a UID for that bidder.

If that's just a special case for these test params, though, then don't worry about it.

File: static/bidder-params/conversant.json [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  "title": "Conversant Adapter Params",
+  "description": "A schema which validates params accepted by the Conversant adapter.",
+  "type": "object",
+  "properties": {
+    "site_id": {
+      "type": "string",
+      "description": "A Conversant specific ID which identifies the site."
+    },
+   	"secure": {
+   	  "type": ["boolean", "integer"],
```
Comment (dbemiller) [inline static/bidder-params/conversant.json]: Now that `FlexBool` is gone, this is only `integer`.

Same goes for `mobile` below.

Comment (pycnvr) [reply static/bidder-params/conversant.json]: Doh!  Thanks.

**Comment by pycnvr** [comment]
I'll like to update SkipNoCookies as well.  Thanks.

**Comment by pycnvr** [comment]
Thanks again for all the help!  Changes have been uploaded, but looks like there is a new requirement for info.yaml.  I could add this too if needed.  Please let me know.

**Comment by pycnvr** [comment]
Info.yaml added as well with mediapsr@conversantmedia.com as the maintainer.  Thanks.

**Comment by pycnvr** [comment]
Awesome.  Thanks again.

**Comment by dbemiller** [comment]
sure thing... just don't forget about OpenRTB :D. Pretty much all of our future development efforts will be going towards that.

**Comment by pycnvr** [comment]
Definitely.  👍

**Pull request merged**